pkg/sharding/apiserver/storage.go

```diff
@@ -0,0 +1,110 @@
+package apiserver
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/apis/meta/internalversion"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/registry/rest"
+	clientrest "k8s.io/client-go/rest"
+)
+
+type storageBase struct {
+	rest.TableConvertor
+
+	// TODO: resolve this information at runtime using an index
+	shards                         map[string]*clientrest.Config
+	shardIdentifierResourceVersion int64
+
+	// we know we will have to delegate to the shards we know about, but the rest.StandardStorage
+	// interface is usually used in a 1:1 mapping to underlying resources, whereas here we want
+	// one of these to handle all request URIs - so, we need to inject information from our callers
+	// to let us make use of the information that would otherwise be lost
+	clientFor  func(cfg *clientrest.Config) (*clientrest.RESTClient, error)
+	requestFor func(client *clientrest.RESTClient, mutators... func(runtime.Object) error) (*clientrest.Request, error)
+}
+
+func (s *storageBase) New() runtime.Object {
+	return &unstructured.Unstructured{}
+}
+
+func (s *storageBase) NewList() runtime.Object {
+	return &unstructured.UnstructuredList{}
+}
+
+func NewMux(base storageBase) *storageMux {
+	return &storageMux{
+		storageBase: base,
+		sharded:     &shardedStorage{storageBase: base},
+		delegating:  &delegatingStorage{storageBase: base},
+	}
+}
+
+type storageMux struct {
+	storageBase
+	sharded listWatcher
+	delegating rest.StandardStorage
+}
+
+func (s *storageMux) Get(ctx context.Context, name string, options *metav1.GetOptions) (runtime.Object, error) {
+	if c := request.ClusterFrom(ctx); c.Wildcard {
```

Comment:
It seems that the master branch points to Kube 1.22 but I didn't find the `ClusterFrom` method in `k8s.io/apiserver/pkg/endpoints/request` package.

Reply:
i suspect the method might be defined in the fork `github.com/kcp-dev/kubernetes/staging/`

Reply:
All I know is it's compiling ... uh... ???

Reply:
(and this predates these changes)